Working on Implementation of Panning

oddtime opened the pull request:
See issue #665

I don't want this pull request to be merged now, but just share it.

In Hydrogen the two pan parameters control two separate balances, like a chain, then if they are hard panned to opposite sides, the signal is killed. If there is a reason to mantain this, or somebody likes it, we can add the new feature as option in preferences

**What is changed** now is,  from a user point of view:
**`note` and `instrument` pan interation**: `note` pan moves the pan in a smaller pan range centered at `instrument `pan.
In this pull req. a resultant pan is calculated then the balance law is called. The signal is never killed.
Of course if the `instrument` pan is set to the left, you cannot move the signal to Hard right with the `note` pan (since  `note` pan range is smaller and centered at `instrument` pan); to achieve this, other strategies would be required as discussed in #665 (for example `note` pan bypassing `instrument` pan, this could be another option in preferences if somebody likes it).

In this pull req, the way hydrogen saves pan is unchanged: pan_L and pan_R for both `note` and ` instrument` objects, even if it is now strange and h2 needs to convert them to single parameters and then applies the pan law again.

I used some pointers to member functions, in the future hypothesis of different pan laws implementation. Just to test them, but they should become song members.

// Edit: Added many pan law options, set in mixer: new button on top-right

oddtime commented:
It is in progress if we decide to add other pan features...
But you can test it and give an opinion on what is already done, described in the first comment

theGreatWhiteShark commented on src/core/Sampler/Sampler.cpp:
```diff
@@ -233,6 +233,29 @@ void Sampler::noteOff(Note* pNote )
+/// functions for pan parameters and laws-----------------
+float Sampler::getRatioPan( float fPan_L, float fPan_R ) {
+	// returns the single pan parameter in [-1,1] from the L,R gains
+	// It doesn't return ERROR if (L,R) = (0,0) nor if they are negative!!!!!
```
Regarding the errors in those two functions: If there is no straight forward default value one can return/set, the C++ way of implementing errors would be an exception. But in Hydrogen it's almost never used (and introducing it would make the code most probably harder to maintain).

I would suggest to introduce these limits to `Instrument::set_pan` and `Note::set_pan` instead. If > 0, `0` will be assigned and a WARNINGLOG triggered and the same for 1 from above.

oddtime replied:
`I would suggest to introduce these limits to Instrument::set_pan and Note::set_pan instead. If > 0, 0 will be assigned and a WARNINGLOG triggered and the same for 1 from above.`

Ok, I keep it for the final version

oddtime replied:
> I would suggest to introduce these limits to Instrument::set_pan and Note::set_pan instead.

There is already a check for negative pan:
```
void Note::set_pan_l( float pan )
{
	__pan_l = check_boundary( pan, PAN_MIN, PAN_MAX );
}

```

In this way H2 one can set both pan_L = pan_R = 0. In this case there is a problem to reconvert back to the pan parameter. It could happen importing a bad song.

theGreatWhiteShark commented on src/core/Sampler/Sampler.cpp:
```diff
@@ -257,6 +280,30 @@ bool Sampler::renderNote( Note* pNote, unsigned nBufferSize, Song* pSong )
+	// TODO a new song member, set in preferences, must point the desired pan parameter Sampler member function
+	float (*getPan)( float, float );
+	// use "ratio" pan parameter (due to the current pan law)
+    getPan = &this->getRatioPan;
```
Why don't you use `getRatioPan` straight away?

oddtime replied:
Not all the pan laws use that definition of the parameter.
In the standard adopted linear panning (a particular case of constant sum pan law i.e. pan_L + pan_R = const) for example the pan parameter has a different dependence on pan_L and pan_R.
So ideally another `getLinearPan( float, float)` function could be pointed

oddtime replied:
You made me think that the implementation could **always** use the method `ratioStraightPolPanLaw(pan)` to save `pan_L` and `pan_R` and **always** `getRatioPan(pan_L, pan_R)` to reconvert them to the initial single parameter. So all this is just a fixed bijection, that we use only to keep unchanged the way h2 stores the pan parameter information. Even if this will be less intuitive for developers.
Then, when rendering the note, the single pan parameter can be interpreted differently depending on the pan law.
So there is no need of `getLinearPan( float, float)` or `getLinearPan( float, float)`.

theGreatWhiteShark commented on src/core/Sampler/Sampler.h:
```diff
@@ -108,7 +108,8 @@ class Sampler : public H2Core::Object
 	 */
 	void reinitializePlaybackTrack();
-
+	static float getRatioPan( float fPan_L, float fPan_R );
```
Could you use Doxygen-style comments to describe the functions? You can e.g. look at the `OscServer` class methods for reference and build the resulting developer documentation using `build.sh d`.

oddtime replied:
Ok

theGreatWhiteShark commented on src/core/Sampler/Sampler.cpp:
```diff
@@ -257,6 +280,30 @@ bool Sampler::renderNote( Note* pNote, unsigned nBufferSize, Song* pSong )
+	float fInstrPan = getPan( pInstr->get_pan_l(), pInstr->get_pan_r() );
+	
+	// resultant pan: note pan moves the pan in a smaller pan range centered at instrument pan
+	float fPan = fInstrPan + fNotePan * ( 1 - fabs( fInstrPan ) );
```
Why do you use a weight for the note pan to make it more pronounced at the center?

oddtime replied:
because the resultant note pan range extension is maximum (really from hard left to hard right) when the instrument pan is at center.
For example when instrument pan is hard panned to the right, there is 0 pan extension for the motion due to note_pan
that was the very meaning of "smaller pan range centered at..."
how small depends on the instrument_pan value.

oddtime replied:
Like a "matrioska" pan
![image](https://user-images.githubusercontent.com/54219881/104101032-98183b00-529d-11eb-8de9-3f6dd3a11e05.png)

theGreatWhiteShark commented on src/core/Sampler/Sampler.cpp:
```diff
@@ -257,6 +280,30 @@ bool Sampler::renderNote( Note* pNote, unsigned nBufferSize, Song* pSong )
+	// use Straight pol pan law.
+	// TODO a new song member, set in preferences, must point the desired pan law Sampler member function		
+	float (*panLaw)( float );
+    panLaw = &this->ratioStraightPolPanLaw;
```
How about making a single `Sampler::panLaw( float, enum)` and an enum class for the different pan laws? Inside this function you could use switch-case statement to distinct between the laws.

oddtime replied:
I preferred to use pointers to bypass the conditions checks and to go straight to the function

oddtime replied:
As the addresses of methods depend on the `Sampler` object, maybe this pointer should be saved in some form in the song (for example written as an `int` or `enum`, which type is new to me now), but should live in `Sample` class.
What do you think?

theGreatWhiteShark replied:
> I preferred to use pointers to bypass the conditions checks and to go straight to the function

I see. But I would like to ask you to use switch-case or if statements nevertheless. It's not because I have a personal reference but the code base of Hydrogen does use the former over function pointers. It's more easy to maintain it if there is a consistent coding style. (Admittedly, the one we deal with is not the best but, well, it works. :wink: )

oddtime replied:
I understand the reason.
I would just ask you if you would consider consistent this alternative:
`float (*m_pPanLaw)( float )` is a member in `Sampler`
While import the song, or setting the preferences, the corresponding member function is assigned to `m_pPanLaw` with `if `statements instead of that weird array of addresses

theGreatWhiteShark commented on src/core/Sampler/Sampler.cpp:
```diff
@@ -257,6 +280,30 @@ bool Sampler::renderNote( Note* pNote, unsigned nBufferSize, Song* pSong )
+	// TODO a new song member, set in preferences, must point the desired pan law Sampler member function		
+	float (*panLaw)( float );
+    panLaw = &this->ratioStraightPolPanLaw;
+    float a = 0.5; // max gain (it has been 0.5 until version 1.0)
```
How about dropping this factor along with [1]

theGreatWhiteShark commented on src/core/Sampler/Sampler.cpp:
```diff
@@ -619,9 +669,10 @@ bool Sampler::renderNote( Note* pNote, unsigned nBufferSize, Song* pSong )
 			cost_L = cost_L * pSong->get_volume();	// song volume
 			cost_L = cost_L * 2; // max pan is 0.5
```
[1] this factor over here? (and the same for the right channel)

oddtime replied:
Good point! I didn't noticed it

trebmuh commented on src/gui/src/PreferencesDialog_UI.ui:
```diff
@@ -564,6 +564,32 @@
+           </item>
+           <item>
+            <property name="text">
+             <string>Costant Power</string>
```
seems to be a typo here (Costant->Co**n**stant)

oddtime replied:
thanks

trebmuh commented on src/gui/src/PreferencesDialog_UI.ui:
```diff
@@ -384,45 +384,45 @@
+              </size>
+             </property>
+             <property name="toolTip">
+              <string>Specifies the variable, which has to remain constant in order to guarantee a working synchronization and relocation in the presence of another Jack timebase master.</string>
```
Since it's an acronym, Jack should be written capitalised: JACK

oddtime replied:
out of the subject of this pull req, I don't know if I can do it right here

oddtime replied:
is it ok?

oddtime replied:
Alsa too is an acronym... I would prefer do not changing these names in the current pull request

trebmuh replied:
I thought it was a new addition since it's a "+" line.
No worries, it can be done later.
_(but still it can be done here while you're at it, there is not potential complication IMHO)_

oddtime replied:
It appeared as if I had added that lines but it was rewritten by the graphical software. However that file is untouched in the last commits.

trebmuh commented on src/gui/src/PreferencesDialog_UI.ui:
```diff
@@ -384,45 +384,45 @@
+              </size>
+             </property>
+             <property name="toolTip">
+              <string>Specifies the variable, which has to remain constant in order to guarantee a working synchronization and relocation in the presence of another Jack timebase master.</string>
```
Since it's an acronym, Jack should be written capitalised: JACK

trebmuh reviewed (changes requested):
Not a programmer, but still helping where I can :wink:

oddtime commented:
I am attaching a songfile to test the pan laws!
It is the same mono sound going **linearly** from left pan to right pan
[panning.zip](https://github.com/hydrogen-music/hydrogen/files/5793211/panning.zip)
Note: **linearly** refers to the pan parameter, not to the perceived angle, which depends on the pan laws.

oddtime commented:
A summary of what is included until the last commit: different pan laws to set in preferences->audio.
Basically at this point you can hear how different laws sound.
Results sound good in my opinion, there's also a huge variety of options (for the pleasure of research), which is not given in any DAW I know.
With the "linear" pan parameter it seems that you have better control of the stereo position with the pan pot, in my opinion.
Ask if you have any questions. I described something in the comments.

The type of pan law is already saved in the songfile, but I am not convinced about the way it is saved.
I am not totally convinced about the way to store the pan law in the core as well (in which class, using indices, pointers etc).

theGreatWhiteShark commented:
>  different pan laws to set in preferences->audio.

> The type of pan law is already saved in the songfile, but I am not convinced about the way it is saved.
I am not totally convinced about the way to store the pan law in the core as well (in which class, using indices, pointers etc).

Hydrogen has two main ways for storing information: within each `.h2song` file and a global `hydrogen.conf` file. The latter should contain all the customization the user is likely to use for a lot of sessions and the former those stuff that might change with every song. The panning law feels like a global setting (especially since it can be set via the preference dialog). Therefore, I would move it from the `Song` class into `Preferences`.

Regarding storing: I think the options the user is able to alter should be in the `Preferences` and everything else can be a member of `Sampler`. Or do you think other parts of the program expect the `Sampler` will use these functions too?

theGreatWhiteShark commented:
>  different pan laws to set in preferences->audio.

> The type of pan law is already saved in the songfile, but I am not convinced about the way it is saved.
I am not totally convinced about the way to store the pan law in the core as well (in which class, using indices, pointers etc).

Hydrogen has two main ways for storing information: within each `.h2song` file and a global `hydrogen.conf` file. The latter should contain all the customization the user is likely to use for a lot of sessions and the former those stuff that might change with every song. The panning law feels like a global setting (especially since it can be set via the preference dialog). Therefore, I would move it from the `Song` class into `Preferences`.

Regarding storing: I think the options the user is able to alter should be in the `Preferences` and everything else can be a member of `Sampler`. Or do you think other parts of the program expect the `Sampler` will use these functions too?

oddtime commented:
> Hydrogen has two main ways for storing information: within each .h2song file and a global hydrogen.conf file. The latter should contain all the customization the user is likely to use for a lot of sessions and the former those stuff that might change with every song. The panning law feels like a global setting (especially since it can be set via the preference dialog). Therefore, I would move it from the Song class into Preferences.

Since the pan law affects how the project sounds and how it will be mixed, it should be stored in the .h2song in my opinion.
We can save it in `hydrogen.conf ` just to keep it saved for the next songs

> Or do you think other parts of the program expect the Sampler will use these functions too?

I don't think so

oddtime commented:
And what about the way of storing in the file? with an integer number? Maybe I would go with a meaningful string

theGreatWhiteShark commented:
> We can save it in hydrogen.conf just to keep it saved for the next songs

I would store it in just one file. Else it will only cause confusion.

> Since the pan law affects how the project sounds and how it will be mixed, it should be stored in the .h2song in my opinion.

I do see your point. But if we store it in the Song file the option must not be set via the `PreferenceDialog`. Do you think user would set different values for each song or just play with it and decide for one that fits them best?

theGreatWhiteShark commented:
> And what about the way of storing in the file? with an integer number? Maybe I would go with a meaningful string

An `enum class`. This way the variable name is a meaningful name but it's value is an integer which is way more convenient to handle. See https://github.com/hydrogen-music/hydrogen/blob/master/src/core/Preferences.h#L375 for an example.

oddtime commented:
> I do see your point. But if we store it in the Song file the option must not be set via the PreferenceDialog. Do you think user would set different values for each song or just play with it and decide for one that fits them best?

I will try to explain why I would like to save it in the song
Actually my concern is: once you change the type of pan law the song sounds in a different way.
The main reason is not only the change of constraint (constant sum, constant power...) indeed, but the way the pan parameter maps the "virtual stereo positions". I mean that the same pan parameter value may "put" the signal in any point of the half stereo panorama depending on the pan law, except for the "central" and "hard-side" positions. We may ask why do not adopt just one bijection between pan and the "virtual stereo placement"? The reason is, in my thought, that there is no a solution to this problem, because the perceived angle could depend on difficult quantities like frequency (also for psycho-acoustics and non-linearity). Also, even if the perceived angle would depend on gain_L and gain_R only, it is not clear what would be the "linear" dependency between gain_L, gain_R and the perceived angle (I said "linear" but of course we don't know any relation). That is to say, the pan laws that I found on manuals, which are all different and inconsistent in my opinion (i.e. switching between them the angle does change, as you can hear experimentally), are all valid theorically. So I included all that pan laws but introduced the "ratio","linear", "quadratic", "polar" distinction to create classes of pan law that map the sounds always in the same points (at least it seems so by ear, and I have another reasonable argument for that...), with compensated gain depending on which pan law. Plus I added a custom pan law which generalises the constraint L^k+R^k=1, making consistent compromises.
All this is to say that if you share a song or for some reason you have changed the pan law setting preference, when you re-open the song you may listen to a very different mix. Tests by ear confirm that (try to switch between "linear" and "quadratic")
On daws also the pan laws are stored in the project (I searched briefly on the web).

> Do you think user would set different values for each song?

If you use pan note property for a surrealistic effect of instruments moving in the space, this could happen. Otherwise I think that every pan law is equivalent, as you will compensate the gain by ear.

theGreatWhiteShark commented:
Yes, you are right. Also when sharing a song the panning law shouldn't be altered by the preferences of the other user.

This still leaves the question of where to put the control/widget to alter this setting. Since it just concerns the internal sound generation it might go to the `Mixer`. But there isn't a place for it yet.

But wait. Does it only concerns the internal sound generation? Doesn't it affect the MIDI export of the notes too? (disclaimer: I have almost no idea about MIDI)

oddtime commented:
> This still leaves the question of where to put the control/widget to alter this setting.

Ok. I approve putting a new widget in the mixer. I think a button on the right, accordingly with the pull req #525 
Another place could be in menu->project

>  Doesn't it affect the MIDI export of the notes too?

In MIDI, pan has its dedicated message from 0-127 so it seems that midi does not transmit the information of separate gain_L and gain_R. I don't know how the pan value is output in midi currently and I should update the code to export the resultant pan coming from instrument and note pan.

oddtime commented:
Here's a space in the mixer:
![Untitled](https://user-images.githubusercontent.com/54219881/104950954-e629ee80-59c1-11eb-9066-0548eebfd024.png)
(maybe too small and bad font)
the button opens a dialog window:
![screenshot-8691da12](https://user-images.githubusercontent.com/54219881/104950965-e88c4880-59c1-11eb-9e4a-97cc09fe0bef.jpg)

theGreatWhiteShark commented:
> Ok. I approve putting a new widget in the mixer. I think a button on the right, accordingly with the pull req #525
Another place could be in menu->project

How about we use a general settings/cog icon in the Mixer instead? This will open a new window from which the pan law dialog can be accessed. This will scale much better with respect to further features/buttons etc. in the mixer.

> is there a strong reason to put it in the mixer?

Since it is only used in the sampler, it only takes effect when using the post fader audio output of Hydrogen. If you use it solely as sequencer to power another drum synth - like drumgizmo - or use the per-track pre fader audio outputs via JACK, the pan law will not alter the result. By placing it in the menu it kinda suggests to affect the panning done in the instrument rack and note property ruler too.

oddtime commented:
Right. I have implemented the cog icon button in the mixer and removed pan law settings from menu.
With that the pan work is almost finished, I just need to review and clean the code.

theGreatWhiteShark commented:
How about introducing 4 headings/categories in the "Select Pan Law" options: linear, polar, ratio, quadratic. Right now it's a little bit hard to get an overview with so much different options densely packed.

oddtime commented:
Well I had the same idea but was lazy to set it

oddtime commented:
Unfortunately I am not an expert of Qt and this seems complex.
Is there a way of putting category headings in a QComboBox, simpler than the following?
https://stackoverflow.com/questions/33012292/grouped-qcombobox

theGreatWhiteShark commented:
> Unfortunately I am not an expert of Qt and this seems complex.
Is there a way of putting category headings in a QComboBox, simpler than the following?
https://stackoverflow.com/questions/33012292/grouped-qcombobox

Well, unfortunately neither am I.

First of all I would recommend using the `QtCreator` (I was manually editing the UI files before realizing this was not the intended way). Doubling clicking the combobox gives some control over its content. Maybe some lines can be disabled and be display as bold or italic. Also, the resolution combobox in the pattern editor features separators. They might help too

oddtime commented:
The current pan law combobox already uses separators but there is not so much contrast to see them clearly.

I used Qt creator to create the dialog, but added the combobox items in `MixerSettingsDialog.cpp` because I can associate the `enum class` to each (which is in `Sampler class`). This seemed good to avoid referring to indices.

oddtime commented:
I found a way

oddtime commented:
Are you ok with the current namings?
Basically _linear_ and _quadratic_ are abbreviations of _"average, linear weights"_ and _"average, quadratic weights"_, _polar_ means _"polar coordinates"_, _ratio_ well means the ratio :)
I hope this makes sense if you have read the comments in the code, otherwise I can explain it better.

theGreatWhiteShark commented:
> I found a way

Nice!

I'll do my best to review the changes during the next days.

> Are you ok with the current namings?
Basically linear and quadratic are abbreviations of "average, linear weights" and "average, quadratic weights", polar means "polar coordinates", ratio well means the ratio :)

Sounds good.

theGreatWhiteShark commented on src/core/Basics/Song.cpp:
```diff
@@ -725,6 +726,53 @@ Song* SongReader::readSong( const QString& sFileName )
+
+	float fPanLawKNorm = LocalFileMng::readXmlFloat( songNode, "pan_law_k_norm", K_NORM_DEFAULT );
+	if ( fPanLawKNorm <= 0.0 ) {
+		fPanLawKNorm = K_NORM_DEFAULT;
```
Could you make `K_NORM_DEFAULT` a static member variable of `Sampler`? This way it will be perfectly clear where to find its definition and documentation.

oddtime replied:
ok

theGreatWhiteShark commented on src/core/LocalFileMgr.cpp:
```diff
@@ -386,6 +386,49 @@ int SongWriter::writeSong( Song * pSong, const QString& filename )
+		sPanLawType = "RATIO_STRAIGHT_POLYGONAL";
+	}
+	LocalFileMng::writeXmlString( songNode, "pan_law", sPanLawType );
+	LocalFileMng::writeXmlString( songNode, "pan_law_k_norm", QString("%1").arg( pSampler->getPanLawKNorm() ) );
```
Could you move this one to `Song` too?

theGreatWhiteShark commented on src/core/Sampler/Sampler.cpp:
```diff
@@ -233,6 +236,229 @@ void Sampler::noteOff(Note* pNote )
+// functions for pan parameters and laws-----------------
+
+float Sampler::getRatioPan( float fPan_L, float fPan_R ) {
+   /** returns the single pan parameter in [-1,1] from the L,R gains
```
Could you move the doc string above the definition in order for Doxygen to find it? You can check using `./build.sh d`. And also in header file to not bloat the source code?

theGreatWhiteShark commented on src/core/Sampler/Sampler.cpp:
```diff
@@ -233,6 +236,229 @@ void Sampler::noteOff(Note* pNote )
+	}
+}
+
+   /** PAN LAWS
```
Nice. Thanks for the thorough documentation! Could you move it above the `enum` definition in the header? Right now, it will be shown as the description of the ` ratioStraightPolygonalPanLaw()` function.

theGreatWhiteShark commented on src/gui/src/Mixer/MixerSettingsDialog_UI.ui:
```diff
@@ -0,0 +1,200 @@
+   </rect>
+  </property>
+  <property name="windowTitle">
+   <string>Form1</string>
```
Could you remove this property? It will be overwritten in the .cpp file but get's listed for the translators who will have no idea what to do with it

theGreatWhiteShark reviewed (changes requested):
Nice. I like it!

Some minor things but almost done.

oddtime commented on src/core/Basics/Song.cpp:
```diff
@@ -725,6 +726,53 @@ Song* SongReader::readSong( const QString& sFileName )
+	
+	// pan law
+	Sampler* pSampler = AudioEngine::get_instance()->get_sampler();
+	QString sPanLawType( LocalFileMng::readXmlString( songNode, "pan_law", "RATIO_STRAIGHT_POLYGONAL" ) );
```
Ok, so how to match?
You mean defining something similar to an array of strings?
`panlawstrings[]`
and referring as `panlawstrings[ RATIO_STRAIGHT_POLYGONAL ]` for example?

theGreatWhiteShark replied:
Since the `enum` can be used as an `int` you can do something like

```c++
int nPanLawType = LocalFileMng::readXmlInt( songNode, "pan_law_type", 0 );
pSong->setPanLawType( nPanLawType );
```

If you use an `enum class` this, however, is not possible and you have to map it's members to `int` or `QString`.

Anyway. It's not that important how to store the variable (as a string or a int) since a) the amount of memory it requires on the system of the user is negligible and b) it's just an internal serialization and not intended to be edited directly. You can also leave it as it is. I don't mind.

oddtime replied:
I see. The reason why I would not save it as an int is because one could add another pan law type in the future (who knows? or for personal pan law tests) and the string keeps the file interpretation save.
I had in mind that writing/reading the string from/to the xml file could have be done in a safer way (I just have to go deeper in the `enum class` )

theGreatWhiteShark replied:
Okay. Then just leave it as is. You don't have to rewrite it as `enum class`. (Why I brought that up the begin with: a number of `enum`s in the code where/are handled like `if ( m_nPanLawType == 5 )` instead of `if ( m_nPanLawType == LINEAR_CONST_SUM )` which makes the code harder to grasp in a totally unnecessary way. But your code hygiene is better, so don't mind :) )

oddtime replied:
Ok!

oddtime commented on src/core/Basics/Song.cpp:
```diff
@@ -725,6 +726,53 @@ Song* SongReader::readSong( const QString& sFileName )
+	Sampler* pSampler = AudioEngine::get_instance()->get_sampler();
+	QString sPanLawType( LocalFileMng::readXmlString( songNode, "pan_law", "RATIO_STRAIGHT_POLYGONAL" ) );
+	if ( sPanLawType == "RATIO_STRAIGHT_POLYGONAL" ) {
+		pSampler->setPanLawType( pSampler->RATIO_STRAIGHT_POLYGONAL );
```
Ok

oddtime commented on src/core/Basics/Song.cpp:
```diff
@@ -725,6 +726,53 @@ Song* SongReader::readSong( const QString& sFileName )
 	pSong->setActionMode( actionMode );
+	
+	// pan law
+	Sampler* pSampler = AudioEngine::get_instance()->get_sampler();
```
Ok

oddtime commented:
Thanks for the check! I did the changes. If you want merge I would add the feature in Changelog

theGreatWhiteShark commented:
> If you want merge I would add the feature in Changelog

Yes, please.

@oddtime one last thing. One of the unit tests does fail. Could you look into it (using `./build.sh t`)?

@trebmuh are all of your requested changes covered? I'm fine with everything as soon as the unit test is resolved.

oddtime commented:
```
1) test: FunctionalTest::testExportAudio (F) line: 130 /home/scarbo/programming/H2Repo/hydrogen/src/tests/FunctionalTests.cpp
Files differ at sample 89898
- Expected: /home/scarbo/programming/H2Repo/hydrogen/src/tests/data//functional/test.ref.flac
- Actual  : /tmp/hydrogen/test-wiTawq.wav
```

Any ideas? Files are 2 s long at 44100Hz...

trebmuh commented:
> @trebmuh are all of your requested changes covered?

They were only typo/cosmetic requests. It looks all good from my PoV.

oddtime commented:
the first tracks are the exported and reference files: test-wiTawq.wav test.ref.flac. The third trck is the difference
![screenshot-6503ea24](https://user-images.githubusercontent.com/54219881/106360469-dba01b00-6318-11eb-97ab-e80ed9fdb08b.jpg)

Normalization shows that the difference is not 0 at this sample.
![Untitled](https://user-images.githubusercontent.com/54219881/106360472-dd69de80-6318-11eb-968a-63bae38690ac.jpg)
but it something like -50dB...
There is another sample like that, may it be caused by a small CPU approximation error (getting back the pan from panL and panR and recalculating the pan law)?

oddtime commented:
I modified the c++ file doing the test to print the different samples in the shell:

`.........................sample1 = 7484, sample2 = 7485`

theGreatWhiteShark commented:
Yeah, two little differences. One in the middle of the song and one at 3/4. I noticed that when diffing the hexdump of these files too.

I did place some prints to `std::cout` in the `DiskWriterDriver` and the `Sampler` to see what's going on. There were quite a number of (very tiny) differences in the float values written in the former. But in `Sampler::renderNote()` the `cost_*` variables have not changed at all. And, AFAIR, this is the very place you did some changes in.

I will keep digging until I find something

oddtime commented:
Thanks @theGreatWhiteShark !

theGreatWhiteShark commented:
Okay. Everything's fine.

There is some truncation happening causing very small differences in the resulting output. In principle we could make the pan calculate use `double` instead of `float` but IMHO this would be over the top.

Could you take one of these temporary results and replace the `test.ref.flac` file?

oddtime commented:
Ok, done.
(but there's still that `test.wav` which seems the old reference)
